.github/workflows/yocto-build-deploy.yml

```diff
@@ -0,0 +1,639 @@
+          path: ${{ runner.temp }}/build-artifacts.tar.zst
+          retention-days: 3
+
+  s3-deploy:
+    name: S3 Deploy
+    runs-on: ${{ fromJSON(inputs.runs-on) }}
+    needs: build
+    if: inputs.deploy-s3 == true
+
+    env:
+      MACHINE: "${{ inputs.machine }}"
+      BALENARC_BALENA_URL: "${{ inputs.balena-environment }}"
+      WORKSPACE: ${{ github.workspace }}
+      VERBOSE: verbose
+
+    defaults:
+      run:
+        working-directory: .
+        shell: bash --noprofile --norc -eo pipefail -x {0}
+
+    steps:
+      # TODO: clone balena-yocto-scripts @ yocto_scripts_ref instead
+      # https://github.com/actions/checkout
+      - name: Clone device repository
+        uses: actions/checkout@v3
+        with:
+          repository: ${{ inputs.device-repo || github.repository }}
+          token: ${{ secrets.GITHUB_TOKEN }}
+          ref: ${{ inputs.device-repo-ref || github.ref }}
+          submodules: true
+
+      - name: Update balena-yocto-scripts submodule to ${{ inputs.yocto-scripts-ref }}
+        if: inputs.yocto-scripts-ref != ''
+        working-directory: ./balena-yocto-scripts
+        run: |
+          git config --add remote.origin.fetch '+refs/pull/*:refs/remotes/origin/pr/*'
+          git fetch --all
+          git checkout --force "${{ inputs.yocto-scripts-ref }}"
+          git submodule update --init --recursive
+
+      # https://github.com/actions/download-artifact
+      - name: Download build artifacts
+        uses: actions/download-artifact@eaceaf801fd36c7dee90939fad912460b18a1ffe # v4.1.2
+        with:
+          path: ${{ runner.temp }}
+          name: build-artifacts
+
+      - name: Extract build artifacts
+        run: |
+          tar -xvf ${{ runner.temp }}/build-artifacts.tar.zst
+          find .
+
+      - name: Set s3 outputs
+        id: s3-lib
+        run: |
+          echo "s3_policy=private" >> $GITHUB_OUTPUT
+          if [ "${{ needs.build.outputs.is_private }}" = "false" ]; then
+            echo "s3_policy=public-read" >> $GITHUB_OUTPUT
+          fi
+
+          echo "deployer_uid=$(id -u)" >> $GITHUB_OUTPUT
+          echo "deployer_gid=$(id -g)" >> $GITHUB_OUTPUT
+
+          s3_bucket_path="${{ inputs.s3-bucket}}/images"
+          if [ "${{ inputs.deploy-esr}}" = "true" ]; then
+            s3_bucket_path="${{ inputs.s3-bucket}}/esr-images"
+          fi
+
+      # https://github.com/docker/setup-buildx-action
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      # login required to pull private balena/balena-img image
+      # https://github.com/docker/login-action
+      - name: Login to Docker Hub
+        uses: docker/login-action@343f7c4344506bcbf9b4de18042ae17996df046d # v3.0.0
+        with:
+          registry: docker.io
+          username: ${{ secrets.DOCKERHUB_USER }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      # https://github.com/docker/build-push-action
+      - name: Build helper image
```

Comment:
Could we not use a pre-built image instead of building for every build?

Reply:
Yes, but then that image would have to be private as well if it is based on balena/balena-img. It's also something we could optimize later.

Also, there is a race condition that while this shared workflow currently lives with balena-yocto-scripts, changes to the helper images would not yet be published while this workflow is being tested. Right now this workflow runs on PRs to balena-yocto-scripts to make sure we haven't broken anything between the scripts and the workflow (deploys will be skipped). But if I change the content of a helper Docker image we would not be testing it unless I make sure this workflow runs after deploying the draft images. Something I can optimize in a future release for sure.

Not an issue for device repos that are using the master branch of balena-yocto-scripts, but still something that needs to be considered if the workflow will always live here.